lib/Server/Grants/OidcAuthCodeGrant.php

```diff
@@ -0,0 +1,411 @@
+<?php
+
+namespace SimpleSAML\Modules\OpenIDConnect\Server\Grants;
+
+use DateTimeImmutable;
+use DateInterval;
+use League\OAuth2\Server\CodeChallengeVerifiers\CodeChallengeVerifierInterface;
+use League\OAuth2\Server\CodeChallengeVerifiers\PlainVerifier;
+use League\OAuth2\Server\CodeChallengeVerifiers\S256Verifier;
+use League\OAuth2\Server\Entities\ClientEntityInterface;
+use League\OAuth2\Server\Entities\UserEntityInterface;
+use League\OAuth2\Server\Exception\OAuthServerException;
+use League\OAuth2\Server\Exception\UniqueTokenIdentifierConstraintViolationException;
+use League\OAuth2\Server\Grant\AuthCodeGrant;
+use League\OAuth2\Server\Repositories\AuthCodeRepositoryInterface;
+use League\OAuth2\Server\Repositories\RefreshTokenRepositoryInterface;
+use League\OAuth2\Server\RequestEvent;
+use League\OAuth2\Server\RequestTypes\AuthorizationRequest;
+use League\OAuth2\Server\ResponseTypes\RedirectResponse;
+use League\OAuth2\Server\ResponseTypes\ResponseTypeInterface;
+use LogicException;
+use Psr\Http\Message\ServerRequestInterface;
+use SimpleSAML\Modules\OpenIDConnect\Entity\Interfaces\OidcAuthCodeEntityInterface;
+use SimpleSAML\Modules\OpenIDConnect\Repositories\Interfaces\OidcAuthCodeRepositoryInterface;
+use SimpleSAML\Modules\OpenIDConnect\Server\Exceptions\OidcServerException;
+use SimpleSAML\Modules\OpenIDConnect\Server\RequestTypes\OidcAuthorizationRequest;
+use SimpleSAML\Modules\OpenIDConnect\Server\ResponseTypes\Interfaces\NonceResponseTypeInterface;
+
+class OidcAuthCodeGrant extends AuthCodeGrant
+{
+    /**
+     * @var DateInterval
+     */
+    protected $authCodeTTL;
+
+    /**
+     * @var CodeChallengeVerifierInterface[]
+     */
+    protected $codeChallengeVerifiers = [];
+
+    /**
+     * @var OidcAuthCodeRepositoryInterface
+     */
+    protected $authCodeRepository;
+
+    public function __construct(
+        AuthCodeRepositoryInterface $authCodeRepository,
+        RefreshTokenRepositoryInterface $refreshTokenRepository,
+        DateInterval $authCodeTTL
+    ) {
+        parent::__construct($authCodeRepository, $refreshTokenRepository, $authCodeTTL);
+
+        $this->authCodeTTL = $authCodeTTL;
+
+        if (\in_array('sha256', \hash_algos(), true)) {
+            $s256Verifier = new S256Verifier();
+            $this->codeChallengeVerifiers[$s256Verifier->getMethod()] = $s256Verifier;
+        }
+
+        $plainVerifier = new PlainVerifier();
+        $this->codeChallengeVerifiers[$plainVerifier->getMethod()] = $plainVerifier;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param ServerRequestInterface $request
+     * @return OidcAuthorizationRequest|AuthorizationRequest
+     * @throws OAuthServerException
+     * @throws OidcServerException
+     */
+    public function validateAuthorizationRequest(ServerRequestInterface $request)
```

Comment:
Yes, they have a common ancestor. I used the docblock @return OidcAuthorizationRequest|AuthorizationRequest to remind me that this method can return those two types. It is like a hint to me, depending if it can support nonce or not. Is this a bad thing to do?

Reply:
I wouldn't say it's a bad thing, but I think it's better to be as strict as reasonably possible..
Right now the method has no return typehint at all, so using the common ancestor as a return type is already an improvement.. I'd leave the docblock as-is, as documentation to tell other developers what this method is supposed to return.

Reply:
That's totally reasonable... Will do! Thank you...